src/passes/OptimizeInstructions.cpp

```diff
@@ -1671,16 +1676,107 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // (x op C1) + (x op C2)   ==>   x * (C1' + C2')
+  // (x op C1) - (x op C2)   ==>   x * (C1' - C2')
+  // where op = (`*`|`<<`)
+  // x * y + x * z   ==>   (y + z) * x
+  // x * y - x * z   ==>   (y - z) * x
+  Expression* optimizeBinaryDistributions(Binary* curr) {
+    using namespace Match;
+    assert(curr->type.isInteger());
+
+    Expression* lhs = nullptr;
+    Expression* rhs = nullptr;
+    Binary* left = nullptr;
+    Binary* right = nullptr;
+    Const* c1 = nullptr;
+    Const* c2 = nullptr;
+
+    if (matches(curr, binary(Abstract::Add, any(&lhs), any(&rhs))) ||
+        matches(curr, binary(Abstract::Sub, any(&lhs), any(&rhs)))) {
+      bool isShlL =
+        matches(lhs, binary(&left, Abstract::Shl, any(), constant(&c1)));
```

Comment:
Try to avoid matching on subexpressions, since that makes it harder to read the full pattern that is being matched. It is better to have a single `matches` expression with a large pattern, even if that pattern gets wrapped onto new lines.